File: api/bot.py
# api/bot.py

import logging
import os
import json
import requests
from telegram import Update, InlineKeyboardButton, InlineKeyboardMarkup
from telegram.ext import Application, CommandHandler, MessageHandler, CallbackQueryHandler, filters, ContextTypes
from telegram.constants import ParseMode

# Enable logging for detailed output. In Vercel, these logs will appear in your deployment logs.
logging.basicConfig(
    format='%(asctime)s - %(name)s - %(levelname)s - %(message)s', level=logging.INFO
)
logger = logging.getLogger(__name__)

# --- Configuration ---
# IMPORTANT: Get your bot token from Vercel Environment Variables (TELEGRAM_BOT_TOKEN)
TELEGRAM_BOT_TOKEN = os.environ.get('TELEGRAM_BOT_TOKEN')
if not TELEGRAM_BOT_TOKEN:
    logger.error("TELEGRAM_BOT_TOKEN environment variable not set.")
    # In a real deployment, you might want to exit or raise an error here.
    # For local testing, you might temporarily hardcode it, but remove for production.

BASE_API_URL = "https://airsongsapi.vercel.app"

# Initialize the Application outside the webhook function to reuse it across invocations
# This is a common pattern for serverless functions with python-telegram-bot
application = Application.builder().token(TELEGRAM_BOT_TOKEN).build()

# --- Helper Functions ---

async def fetch_songs_from_api(query: str):
    """Fetches song data from the JioSaavn API."""
    try:
        url = f"{BASE_API_URL}/result/?query={requests.utils.quote(query)}"
        logger.info(f"Fetching songs from: {url}")
        response = requests.get(url)
        response.raise_for_status()  # Raise an HTTPError for bad responses (4xx or 5xx)
        return response.json()
    except requests.exceptions.RequestException as e:
        logger.error(f"Error fetching songs from API: {e}")
        return None

async def fetch_lyrics_from_api(song_id: str):
    """Fetches lyrics for a given song ID from the JioSaavn API."""
    try:
        url = f"{BASE_API_URL}/lyrics/?query={requests.utils.quote(song_id)}"
        logger.info(f"Fetching lyrics from: {url}")
        response = requests.get(url)
        response.raise_for_status()
        return response.json()
    except requests.exceptions.RequestException as e:
        logger.error(f"Error fetching lyrics from API: {e}")
        return None

# --- Telegram Bot Handlers ---

async def start_command(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handles the /start command."""
    await update.message.reply_text(
        "Hello! I am QuantXsongs Bot. Send me a song name to search for it, "
        "and I'll provide streaming links, download options, and lyrics."
    )

async def search_songs(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Searches for songs based on user's text message."""
    query = update.message.text
    logger.info(f"Received search query: '{query}' from user {update.effective_user.id}")

    await update.message.reply_text(f"Searching for '{query}'... Please wait, this might take a moment.")

    songs = await fetch_songs_from_api(query)

    if not songs:
        await update.message.reply_text("Sorry, I couldn't fetch any songs at the moment. Please try again later.")
        return

    if not songs: # This check is redundant if the previous one handles None, but good for empty list
        await update.message.reply_text(f"No songs found for '{query}'. Please try a different search term.")
        return

    # Limit to a reasonable number of results to avoid spamming chat
    for song in songs[:5]: # Send top 5 results
        if song.get('disabled') == "true":
            status_text = song.get('disabled_text', 'Pro Only')
            message_text = (
                f"🎶 *{song.get('song', 'N/A')}* by {song.get('primary_artists', 'N/A')}\n"
                f"💿 Album: {song.get('album', 'N/A')}\n"
                f"⏳ Duration: {int(song.get('duration', 0)) // 60}:"
                f"{int(song.get('duration', 0)) % 60:02d}\n"
                f"🚫 Status: {status_text} (Streaming/Download not available)"
            )
            await update.message.reply_text(message_text, parse_mode=ParseMode.MARKDOWN)
            continue

        audio_url = song.get('media_url') or song.get('media_preview_url') # Prefer full URL
        image_url = song.get('image')
        perma_url = song.get('perma_url')

        if not audio_url:
            message_text = (
                f"🎶 *{song.get('song', 'N/A')}* by {song.get('primary_artists', 'N/A')}\n"
                f"💿 Album: {song.get('album', 'N/A')}\n"
                f"⏳ Duration: {int(song.get('duration', 0)) // 60}:"
                f"{int(song.get('duration', 0)) % 60:02d}\n"
                f"⚠️ No playable audio found for this song."
            )
            await update.message.reply_text(message_text, parse_mode=ParseMode.MARKDOWN)
            continue

        # Create inline keyboard buttons
        keyboard = []
        if song.get('id'):
            # Callback data for lyrics: 'lyrics_<song_id>'
            keyboard.append(InlineKeyboardButton("Get Lyrics", callback_data=f"lyrics_{song.get('id')}"))
        if perma_url:
            # URL button for JioSaavn link
            keyboard.append(InlineKeyboardButton("View on JioSaavn", url=perma_url))

        reply_markup = InlineKeyboardMarkup([keyboard]) if keyboard else None

        caption_text = (
            f"🎶 *{song.get('song', 'N/A')}* by {song.get('primary_artists', 'N/A')}\n"
            f"💿 Album: {song.get('album', 'N/A')}\n"
            f"⏳ Duration: {int(song.get('duration', 0)) // 60}:"
            f"{int(song.get('duration', 0)) % 60:02d}"
        )

        try:
            # Send audio with thumbnail if available
            await update.message.reply_audio(
                audio=audio_url,
                caption=caption_text,
                parse_mode=ParseMode.MARKDOWN,
                thumbnail=image_url, # Telegram can fetch thumbnail from URL
                performer=song.get('primary_artists', 'N/A'),
                title=song.get('song', 'N/A'),
                reply_markup=reply_markup
            )
        except Exception as e:
            logger.error(f"Failed to send audio for {song.get('song')}: {e}")
            # Fallback to sending text message if audio sending fails
            await update.message.reply_text(
                f"🎶 *{song.get('song', 'N/A')}* by {song.get('primary_artists', 'N/A')}\n"
                f"💿 Album: {song.get('album', 'N/A')}\n"
                f"⏳ Duration: {int(song.get('duration', 0)) // 60}:"
                f"{int(song.get('duration', 0)) % 60:02d}\n"
                f"🔗 [Listen/Download Here]({audio_url})\n"
                f"⚠️ Could not send audio directly, but here's the link. "
                f"Status: {song.get('disabled_text', 'Available')}",
                parse_mode=ParseMode.MARKDOWN,
                reply_markup=reply_markup
            )

async def button_callback_handler(update: Update, context: ContextTypes.DEFAULT_TYPE):
    """Handles inline keyboard button presses."""
    query = update.callback_query
    await query.answer() # Acknowledge the callback query to remove the loading spinner

    data = query.data
    logger.info(f"Received callback data: {data}")

    if data.startswith('lyrics_'):
        song_id = data.split('_')[1]
        # Update the message to show "Fetching lyrics..."
        await query.edit_message_text(text="Fetching lyrics... Please wait.")

        lyrics_data = await fetch_lyrics_from_api(song_id)

        if lyrics_data and lyrics_data.get('lyrics'):
            lyrics_text = lyrics_data['lyrics']
            # Telegram has a message length limit (4096 characters), truncate if necessary
            if len(lyrics_text) > 4096:
                lyrics_text = lyrics_text[:4000] + "\n\n... (lyrics truncated due to Telegram message limit)"
            await query.edit_message_text(text=f"📜 *Lyrics:*\n\n{lyrics_text}", parse_mode=ParseMode.MARKDOWN)
        else:
            await query.edit_message_text(text="Sorry, lyrics not found for this song.")
    else:
        await query.edit_message_text(text="Unknown action.")

# --- Setup for Webhook ---

# Add handlers to the application
application.add_handler(CommandHandler("start", start_command))
application.add_handler(MessageHandler(filters.TEXT & ~filters.COMMAND, search_songs))
application.add_handler(CallbackQueryHandler(button_callback_handler))

# The Vercel entry point for the serverless function
async def handler(request):
    if request.method == "POST":
        # Telegram sends updates as JSON in the request body
        body = await request.get_json()
        update = Update.de_json(body, application.bot)
        await application.process_update(update)
        return {"statusCode": 200, "body": "OK"}
    elif request.method == "GET":
        # For initial webhook setup or health check
        return {"statusCode": 200, "body": "Bot is running!"}
    else:
        return {"statusCode": 405, "body": "Method Not Allowed"}

File: bot.js
const TelegramBot = require('node-telegram-bot-api');
const axios = require('axios');

const token = process.env.TELEGRAM_BOT_TOKEN;
if (!token) {
  console.error('❌ TELEGRAM_BOT_TOKEN environment variable is not set!');
  process.exit(1);
}

const bot = new TelegramBot(token, { polling: true });
const API_BASE_URL = 'https://airsongsapi.vercel.app';

console.log('🤖 AirSongs Telegram Bot is running (polling mode)...');

// /start command
bot.onText(/\/start/, (msg) => {
  bot.sendMessage(msg.chat.id, `
🎵 *Welcome to AirSongs Bot!* 🎵

Search for any song and I'll help you:
• 🎧 Stream music directly
• 📥 Download MP3 files
• 📝 Get lyrics
• ℹ️ View song details

Just type the name of any song to get started!

*Examples:*
• Arjan Vailly
• Shape of You
• Blinding Lights

Built with ❤️ by AirSongs
  `, { parse_mode: 'Markdown' });
});

// /help command
bot.onText(/\/help/, (msg) => {
  bot.sendMessage(msg.chat.id, `
🤖 *AirSongs Bot Commands:*

/start - Start the bot
/help - Show this help message

🔍 *How to use:*
1. Send me any song name
2. Choose from the search results
3. Stream, download, or get lyrics!

💡 *Tips:*
• Be specific with song names for better results
• Include artist name for more accurate search
• All downloads are in high quality MP3 format

🎵 Enjoy your music!
  `, { parse_mode: 'Markdown' });
});

// Song search
bot.on('message', async (msg) => {
  const chatId = msg.chat.id;
  const messageText = msg.text;

  if (!messageText || messageText.startsWith('/')) return;

  try {
    bot.sendChatAction(chatId, 'typing');

    const response = await axios.get(`${API_BASE_URL}/result/?query=${encodeURIComponent(messageText)}`);

    if (!Array.isArray(response.data) || response.data.length === 0) {
      return bot.sendMessage(chatId, '❌ No songs found. Try a different search term.');
    }

    const songs = response.data.slice(0, 5);
    await bot.sendMessage(chatId, `🔍 Found ${songs.length} results for "${messageText}":`);

    for (const song of songs) {
      const duration = `${Math.floor(song.duration / 60)}:${String(song.duration % 60).padStart(2, '0')}`;
      const songInfo = `🎵 *${song.song}*\n👤 Artist: ${song.primary_artists}\n💽 Album: ${song.album}\n⏱️ Duration: ${duration}\n🗓️ Year: ${song.year}\n🌐 Language: ${song.language}`;

      const keyboard = {
        inline_keyboard: [
          [
            { text: '🎧 Stream', callback_data: `stream_${song.id}` },
            { text: '📥 Download', callback_data: `download_${song.id}` }
          ],
          [
            { text: '📝 Lyrics', callback_data: `lyrics_${song.id}` },
            { text: 'ℹ️ Info', callback_data: `info_${song.id}` }
          ]
        ]
      };

      if (song.image) {
        await bot.sendPhoto(chatId, song.image, {
          caption: songInfo,
          parse_mode: 'Markdown',
          reply_markup: keyboard
        });
      } else {
        await bot.sendMessage(chatId, songInfo, {
          parse_mode: 'Markdown',
          reply_markup: keyboard
        });
      }
    }
  } catch (error) {
    console.error('Search error:', error.message);
    bot.sendMessage(chatId, '❌ Sorry, there was an error searching for songs. Please try again.');
  }
});

// Button press handler
bot.on('callback_query', async (callbackQuery) => {
  const chatId = callbackQuery.message.chat.id;
  const data = callbackQuery.data;

  try {
    const underscoreIndex = data.indexOf('_');
    const action = data.substring(0, underscoreIndex);
    const songId = data.substring(underscoreIndex + 1);

    const songResponse = await axios.get(`${API_BASE_URL}/song/?query=${songId}`);
    let song;
    if (Array.isArray(songResponse.data) && songResponse.data.length > 0) {
      song = songResponse.data[0];
    } else {
      return bot.answerCallbackQuery(callbackQuery.id, { text: '❌ Song not found!' });
    }

    switch (action) {
      case 'stream':
        bot.sendChatAction(chatId, 'upload_audio');
        if (song.media_url) {
          await bot.sendAudio(chatId, song.media_url, {
            title: song.song,
            performer: song.primary_artists,
            duration: parseInt(song.duration)
          });
          bot.answerCallbackQuery(callbackQuery.id, { text: '🎧 Streaming...' });
        } else {
          bot.answerCallbackQuery(callbackQuery.id, { text: '❌ Stream not available!' });
        }
        break;

      case 'download':
        if (song.media_url) {
          await bot.sendMessage(chatId, `📥 *Download Link:*\n${song.media_url}\n\n💡 Click the link to download the MP3 file.`, { parse_mode: 'Markdown' });
          bot.answerCallbackQuery(callbackQuery.id, { text: '📥 Download link sent!' });
        } else {
          bot.answerCallbackQuery(callbackQuery.id, { text: '❌ Download not available!' });
        }
        break;

      case 'lyrics':
        bot.sendChatAction(chatId, 'typing');
        try {
          const lyricsResponse = await axios.get(`${API_BASE_URL}/lyrics/?query=${songId}`);
          if (lyricsResponse.data.success && lyricsResponse.data.data?.lyrics) {
            const lyrics = lyricsResponse.data.data.lyrics;
            const truncated = lyrics.length > 3800 ? lyrics.substring(0, 3800) + '\n...' : lyrics;
            await bot.sendMessage(chatId, `📝 *Lyrics for ${song.song}*\n\n${truncated}`, { parse_mode: 'Markdown' });
            bot.answerCallbackQuery(callbackQuery.id, { text: '📝 Lyrics loaded!' });
          } else {
            bot.sendMessage(chatId, '❌ Lyrics not available for this song.');
            bot.answerCallbackQuery(callbackQuery.id, { text: '❌ No lyrics found!' });
          }
        } catch {
          bot.sendMessage(chatId, '❌ Error fetching lyrics.');
          bot.answerCallbackQuery(callbackQuery.id, { text: '❌ Error fetching lyrics!' });
        }
        break;

      case 'info': {
        const duration = `${Math.floor(song.duration / 60)}:${String(song.duration % 60).padStart(2, '0')}`;
        await bot.sendMessage(chatId, `ℹ️ *Song Information*\n\n🎵 *Title:* ${song.song}\n👤 *Artist:* ${song.primary_artists}\n💽 *Album:* ${song.album}\n⏱️ *Duration:* ${duration}\n🗓️ *Year:* ${song.year}\n🌐 *Language:* ${song.language}\n▶️ *Play Count:* ${song.play_count ? parseInt(song.play_count).toLocaleString() : 'N/A'}\n🏷️ *Label:* ${song.label || 'N/A'}`, { parse_mode: 'Markdown' });
        bot.answerCallbackQuery(callbackQuery.id, { text: 'ℹ️ Song info displayed!' });
        break;
      }

      default:
        bot.answerCallbackQuery(callbackQuery.id, { text: '❌ Unknown action!' });
    }
  } catch (error) {
    console.error('Callback error:', error.message);
    bot.answerCallbackQuery(callbackQuery.id, { text: '❌ Error processing request!' });
  }
});

bot.on('polling_error', (error) => {
  console.error('Polling error:', error.message);
});
